Fail steps/tests/suites when expressions throw exceptions

jonsmock opened the pull request:
This PR conceptually adds try/catch around expressions, failing the step/test/suite if an expression throws an exception.  Also, it fixes a bug, where if an `expect` condition fails, the `run` is not retried even if `repeat` is set correctly.

Before we add catching of expression exceptions, there is a pretty substantial refactoring to move the run-(step|test|suite) functions to become "pipelines" using `->` and `while->`.  `while->` is a custom macro that takes a predicate as its first argument, then threads the initial value through functions as long as the previous initial or returned value matches the predicate.  This is used as an "early exit" mechanism, as we interpolate various parts of steps/tests/suites.

Misc:

- Dockerfile cleanup
- Verifying skipped test count in GHA when omitting `--continue-on-error`
- Interpolating step/test/suite names

jonsmock committed:
Fix FromAsCasing in Dockerfile

https://docs.docker.com/reference/build-checks/from-as-casing/

jonsmock committed:
Use docker compose with newer plugin CLI syntax

docker-compose no longer supported in GHA

jonsmock committed:
Repeat/re-run exec when 'expect' assertions fail

The 'expect' assertions are conceptually part of the 'run' success; if
'expect' assertions fail, the 'run' should be retried.

We do not need to re-evaluate the 'env' and other keys that are
evaluated before the 'run' executes; however, we do need to re-execute
the 'run' and re-evaluate the 'expect' assertions.

gdw2vs commented on src/dctest/core.cljs:
```diff
@@ -202,21 +207,19 @@ Options:
+                            (P/let [results (run-exec step-context step)
+                                    step-context (assoc step-context :step results)
+                                    error (or (:error results)
+                                              (first (keep #(get-expect-error step-context %) (:expect step))))]
```
why only the first one?

jonsmock replied:
I intend to return all, but I was trying to keep the refactoring/diff smaller for this PR.

gdw2vs commented on src/dctest/core.cljs:
```diff
@@ -177,87 +179,119 @@ Options:
-  (when-not (expr/read-eval context expect)
-    {:message (str "Expectation failure: " expect)
-     :debug (expr/explain-refs context expect)}))
+(defn update-any-with-context [{:keys [context results]} k f]
```
I'd change `k` to be a path, so down on line 224, instead of calling:

`(read-eval-key :if)`

you'd call:

`(read-eval-key [:results :if])`

`read-eval-key` seems too genericily-named to require the reader or caller to know that the passed-in `k` is in reference to the `:results` when the whole context/results map is being passed in.

gdw2vs replied:
What's the `any` meant to convey in `update-any-with-context`?

jonsmock replied:
The `k` vs `ks` thing was one of the things I thrashed on quite a bit as I worked through options here.  Ultimately I went this direction for two reasons:  1) all of the functions here take and return context+results essentially, so I thought it might be ok that results was not explicit (and perhaps even weird if it were explicit), and 2) I didn't update error messages yet (I had implemented it a few times, but again the diff was larger and harder to process), but I would like good, localized error messages, ex: "Error evaluating 'if': ...".  It seemed weird to me if the function that created that message ignored the first key in `ks` if we pass `[:results :if]`.

But I do see what you're saying. Do you think a better name for the function would help, or is this just misconceived?

As for `update-any-with-context`, yeah naming was hard here again.  This function is like `update` except it carries/lifts the context along as the first argument.  Also, it looks at the type of the thing it's updating (is it a string? vector? map?), and handles those cases slightly differently.  We don't mind if our values are in a "container" like vector or map, we just want to update them all.  The `any` is trying to differentiate this aspect from the normal `update` behavior, I guess (think "any type"?).  Suggestions definitely welcome.

jonsmock replied:
Nevermind - this can be cleaner, I'll push changes. Thanks!

jonsmock replied:
Pushed a change with a more generic `update-in*` (better names welcome)

gdw2vs commented on src/dctest/core.cljs:
```diff
@@ -177,87 +179,119 @@ Options:
+
+                 (interpolate-key :env)
+                 (as-> m
+                   (update-in m [:context :env] merge (get-in m [:results :env])))
```
Add a comment on _why_?

jonsmock replied:
All expressions in suites/test/steps are evaluated using the current context.  For instance, if you see `${{ env.FOO }}` the `env` is coming from the context `:env` map.  So any time a suite/test/step has env settings, we need to resolve them (in this case using `interpolate-key`) and then merge the resolved values back into the context env.  Perhaps there's a clearer way to do this?

gdw2vs replied:
From just looking at this function, I'm unable to tell whether it's important that the interpolated env be in both the step (result) and the context because it's put in one place and then copied to the other.  So is it needed in both?  To answer this question, I need to dig deeper.  I think it only needs to be in context.  

So could these two expressions be replaced by:

```clojure
(as-> m
   (update-in m [:context :env] (interpolate-key :env))
```

I know it can't, b/c the present `interpolate-key` potentially does more than just interpolate the key (e.g. in the case of error).  I guess the moral is mixing monads and non-monads is hard.

jonsmock replied:
This is somewhat related to the discussion in the other thread, where we have to "pop" env off of the context, when we're done.  I think it would actually be cleaner if the interpolation _only_ happened in the suite/test/step, and then when we go to read-eval/interpolate, we combine the env's from that stack at the last second.  There are some details/challenges there that seemed to make this path messy, but perhaps I should give that another shot.

gdw2vs commented on src/dctest/core.cljs:
```diff
@@ -177,87 +179,119 @@ Options:
+
+                 (assoc-in [:results :outcome] :pass))
+
+        (assoc :context orig-context)
```
I think there's sort of a presumption in a thread macro that you're going to build up something and then return it.  I'm surprised to see you throw away any changed context here!

(thinks for a minute....)

I think this is an indication that they're probably different contexts and, if so, I'd probably call them different things.  Maybe at the top you could have:

```clojure
(P/-> {:context orig-context
       :working-context orig-context
       :results (assoc step :outcome :pending)}
```

But then, of course, you'd need to discard `:working-context` at the end.  🤷 

(thinks for another minute...)

It sort of clicked why you don't want to 'return' the modified state.   Each step should be handed the same context and one step shouldn't be able to change the context for a sibling step.  If that's the case, then I think the responsibility of ensuring that a sibling can't mess with another sibling's context should be on the caller, not on this function.  

(thinks more harder...)

A step shouldn't be able to modify the context for siblings or parents... except in the case of `[:context :state :failed]`.  I guess I'd consider moving line 243, below, up to the parent function so that the caller can ignore any returned context and pass the same un-mutated context to each step.

gdw2vs replied:
`orig-context` is immutable.  Maybe just `context`?

jonsmock replied:
This was the other area I thrashed on a lot.  I agree that the ideal would be the context is just threaded straight through.  There are definitely cases were tests or steps _do_ want to modify the context for future tests/steps.  The first one is to update the `[:context :state :failed]`, yup (and this could also be rethought).  The second one that we haven't added yet is that jobs and steps in GHA provide outputs that future jobs and steps can refer to.  These outputs would go into the context in our case.

The thing we need to be careful of at the moment is that any env in the current suite/test/step is not leaked into future suites/tests/steps.  Perhaps this is the only thing we actually "revert" here.  I considered having a stack (a vector) of envs that gets pushed to and popped off, but I didn't actually implement it to see if it was as weird as I expected it to be.  I'll try that.

jonsmock replied:
Addressed!  context is now only passed down, never "mutated" and returned back "up"

gdw2vs commented on src/dctest/expressions.cljs:
```diff
@@ -284,7 +287,16 @@ ExpectedInterpolation ::= InterpolatedExpression PrintableChar*
+                                     {:error-msg (.-message err)}))]
+                           (if error-msg
+                             (str "<Error: " error-msg ">")
+                             result)))]
```
Would this be work?  If so, is it better?

```clojure
(fn [context expr]
  (try
    (read-eval context expr)}
    (catch js/Error err
      (str "<Error: " (.-message error-msg) ">")))))
```

jonsmock replied:
Applied, thanks!

gdw2vs commented on src/dctest/expressions.cljs:
```diff
@@ -284,7 +287,16 @@ ExpectedInterpolation ::= InterpolatedExpression PrintableChar*
+                           (if error-msg
+                             (str "<Error: " error-msg ">")
+                             result)))]
+    (reduce #(assoc %1 %2 (safe-read-eval context %2))
```
Are `%1` and `%2` always maps that can be `assoc`d?  In the case of an error, isn't it just a string?

jonsmock replied:
`%1` is a map (expressions to values/results) that we're building up, and `%2` is the expression we are evaluating (which could result in a string, exactly).

jonsmock committed:
Simplify:  stop returning "mutated" contexts

These test/step execution functions were returning the context for two
reasons:

* they modified the context's :state :failed boolean (which somewhat
  conflates the overall run state and the individual test state)
* they anticipated supporting test and step outputs in the future

However, the :state :failed boolean would then have to be reset to false
in 'run-test, if the user was using --continue-on-error. Additionally,
returning the context meant we had to restore the previous :env.

Instead, it's much simpler to treat the context as immutable! If we need
to support outputs in the future, they should be returned as part of the
test/step result, and the higher level function can determine how to add
those outputs to a newly created context for the next test/step
execution.

jonsmock committed:
Normalize step name, move expect/index into 'normalize

jonsmock committed:
Interpolate suite/test/step names, add 'interpolate-any

jonsmock committed:
Extract loop/reduce functions

Extracts 'run-expectations, 'execute-step-retries, 'run-tests, and
'run-suites. The first three of these return a step, step, and suite
respectively, so that the bodies of the 'run-test and 'run-suite
functions can use them with clojure.core/update.

Reorganize logging a bit.

jonsmock committed:
Use pending-> to catch and short-circuit on exceptions

Expressions in suite/test/step can throw, in which case we want to stop
evaluating and fail the suite/test/step (ideally preserving as much
information as possible, such as the interpolated :name). This change
adds a pending-> macro to handle the common pattern of:

1. check if the suite/test/step is skipped or failed (if so, short
   circuit)
2. otherwise, attempt to do the next thing (ex: evaluate an env, run a
   test's steps, etc)
3. if an unexpected exception is thrown, catch it, and mark the
   suite/test/step as :fail with an error

Helpers pass!/skip!/fail! are added to manipulate the suite/test/step
outcome consistently and tersely.

The pending-> macro has to be extracted into its own namespace and
duplicated in both clj and cljs files to support ShadowCLJS and nbb.

jonsmock committed:
Show errors from all levels (suite, test, or step)

Now that suites and tests can throw exceptions, we switch to recording
the error on the suite, test, or step as appropriate (previously the
step errors were bubbled up and also assoc'd to the test).

To support this, we summarize errors at all levels in our summary before
displaying the list to the user. As an improvement, we also show where
the error came from (previously the user had to infer which step caused
the failure).

jonsmock commented:
Main change from original description:

`while->` is discarded in favor of `pending->` which acts like `->`, implicitly `try/catch`s each form, and only proceeds to evaluate the next form if the initial/resulting expression is `pending?`. This allows us to take a suite/test/step and evaluate/interpolate different parts of them until an error occurs, at which point we short-circuit the evaluation and return a failure (retaining any parts we _did_ evaluate sucessfully, such as the `:name`)

kanaka commented on src/dctest/outcome.clj:
```diff
@@ -0,0 +1,24 @@
+                                                      {:message (str "Exception thrown: " (.-message ~err))}))))
+                   forms)]
+    `(promesa.core/let [~sts ~suite-test-or-step
+                        ~@(interleave (repeat sts) (butlast forms))]
```
Nice. I understand now what this is doing, but it took some pondering. A comment would be worthwhile.

kanaka commented on src/dctest/core.cljs:
```diff
@@ -170,91 +175,91 @@ Options:
+  (P/let [step (pending-> step
+                          (->> (skip-if-necessary context))
+                          (update :env #(interpolate-any context %)))
+          context (update context :env merge
```
The way the context manipulation interrupts the `pending->` flow makes me wonder if this points to some higher abstraction. Just a thought.

jonsmock replied:
Believe me, I've been thinking about that nonstop.  If you have even a sketch of a better idea, I'd be willing to try to simplify this more.  This was something like my 4th or 5th attempt.  I think my ideal would be something where the env doesn't need to be copied into the context, but rather, simply by interpolating it in the suite/test/step, it now participates in future interpolating ... but I couldn't figure out how to make that work that wasn't more complicated than this.

kanaka replied:
Yeah, if I come up with something I'll let you know ;-). Maybe after some hammock time.

kanaka commented on src/dctest/core.cljs:
```diff
@@ -363,14 +408,17 @@ Options:
         (* 1000))))
 
 (defn normalize [suite path]
```
A brief docstring summarizing what sort of normalizing we are doing would be good.

kanaka reviewed (approved):
We could use more docstrings and inline comments. It's pretty dense code with a reasonable level of cyclomatic complexity.

I've marked it approved because I think the doc improvement is general rather than something specific to this change. There is some great information in the commits that should become inline comments/docstrings.

jonsmock commented:
Thanks, everyone, for the very helpful reviews and feedback.  I'll definitely address the docstring/comments in a separate PR.
